cpp/src/neighbors/detail/hnsw.hpp

```diff
@@ -206,6 +206,213 @@ std::enable_if_t<hierarchy == HnswHierarchy::CPU, std::unique_ptr<index<T>>> fro
 }
 
+int initialize_point_in_hnsw(hnswlib::HierarchicalNSW<float>* appr_algo,
+                             raft::host_matrix_view<const float, int64_t, raft::row_major> dataset,
+                             int64_t real_index,
+                             int curlevel)
+{
+  auto cur_c                        = appr_algo->cur_element_count++;
+  appr_algo->element_levels_[cur_c] = curlevel;
+  memset(appr_algo->data_level0_memory_ + cur_c * appr_algo->size_data_per_element_ +
+           appr_algo->offsetLevel0_,
+         0,
+         appr_algo->size_data_per_element_);
+
+  // Initialisation of the data and label
+  memcpy(appr_algo->getExternalLabeLp(cur_c), &real_index, sizeof(hnswlib::labeltype));
+  memcpy(appr_algo->getDataByInternalId(cur_c),
+         dataset.data_handle() + real_index * dataset.extent(1),
+         appr_algo->data_size_);
+
+  if (curlevel) {
+    appr_algo->linkLists_[cur_c] = (char*)malloc(appr_algo->size_links_per_element_ * curlevel + 1);
+    if (appr_algo->linkLists_[cur_c] == nullptr)
+      throw std::runtime_error("Not enough memory: addPoint failed to allocate linklist");
+    memset(appr_algo->linkLists_[cur_c], 0, appr_algo->size_links_per_element_ * curlevel + 1);
+  }
+  return cur_c;
+}
+
+template <typename T, HnswHierarchy hierarchy>
+std::enable_if_t<hierarchy == HnswHierarchy::GPU, std::unique_ptr<index<T>>> from_cagra(
+  raft::resources const& res,
+  const index_params& params,
+  const cuvs::neighbors::cagra::index<T, uint32_t>& cagra_index,
+  std::optional<raft::host_matrix_view<const T, int64_t, raft::row_major>> dataset)
+{
+  auto host_dataset = raft::make_host_matrix<T, int64_t>(0, 0);
+  raft::host_matrix_view<const T, int64_t, raft::row_major> host_dataset_view(
+    host_dataset.data_handle(), host_dataset.extent(0), host_dataset.extent(1));
+  if (dataset.has_value()) {
+    host_dataset_view = dataset.value();
+  } else {
+    // move dataset to host, remove padding
+    auto cagra_dataset = cagra_index.dataset();
+    host_dataset =
+      raft::make_host_matrix<T, int64_t>(cagra_dataset.extent(0), cagra_dataset.extent(1));
+    RAFT_CUDA_TRY(cudaMemcpy2DAsync(host_dataset.data_handle(),
+                                    sizeof(T) * host_dataset.extent(1),
+                                    cagra_dataset.data_handle(),
+                                    sizeof(T) * cagra_dataset.stride(0),
+                                    sizeof(T) * host_dataset.extent(1),
+                                    cagra_dataset.extent(0),
+                                    cudaMemcpyDefault,
+                                    raft::resource::get_cuda_stream(res)));
+    raft::resource::sync_stream(res);
+    host_dataset_view = host_dataset.view();
+  }
+
+  // initialize hnsw index
+  auto hnsw_index =
+    std::make_unique<index_impl<T>>(cagra_index.dim(), cagra_index.metric(), hierarchy);
+  auto appr_algo = std::make_unique<hnswlib::HierarchicalNSW<typename hnsw_dist_t<T>::type>>(
```

Comment:
Why do we need to use hnsw_dist_t<T> type? Could you add a note?

I tested conversion using `float` and `uist8_t` datasets, and it works (after applying my suggestions).